tests/test_general_decompositions.py

```diff
@@ -0,0 +1,70 @@
+#       this list of conditions and the following disclaimer.
+#
+#    2. Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#
+#    3. Neither the name of the QuTiP: Quantum Toolbox in Python nor the names
+#       of its contributors may be used to endorse or promote products derived
+#       from this software without specific prior written permission.
+#
+#    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+#    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+#    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+###############################################################################
+import numpy as np
+import pytest
+from qutip_qip.decompositions.general_decompositions import normalize_matrix, check_unitary
+from qutip import Qobj
+
+def test_normalize_matrix_valid_input_not_unitary():
+    """ Test output when a valid input is provided which is not a unitary.
+    """
+    normalized_M = normalize_matrix(np.array([[1, 2], [3, 4]]))
+    calculated_M = np.array([[0.70711, 1.41421],[2.12132, 2.82843]])
+    assert np.array_equal(calculated_M, normalized_M)
+    assert check_unitary(calculated_M) == False
+
+@pytest.mark.parametrize("invalid_input", (Qobj([[1],[2],[3],[4],[5]]),(1,2)))
+def test_normalize_matrix_invalid_input(invalid_input):
+    """ Test error when Qobj array is provided as input.
+    """
+    with pytest.raises(TypeError, match="Not a valid input : A Numpy input array must be provided."):
+        normalize_matrix(invalid_input)
+
+def test_normalize_matrix_empty_array():
+    """When an empty array is provided as input."""
+    with pytest.raises(ValueError, match="An empty array was provided as input."):
+        normalize_matrix(np.array([]))
+
+
+def test_normalize_matrix_non_square_matrix():
+    """Test error is raised when row number and column number of Matrix
+    are mismatched."""
+    with pytest.raises(ValueError, match="Input must be a square matrix to be a valid gate."):
+                    normalize_matrix(np.array([[1,4,3]]))
+                    normalize_matrix(np.array([[1]]))
+                    normalize_matrix(np.array([[1, 2], [3, 4],[5,6]]))
```

Comment:
There is an issue with using `@pytest.mark.parametrize` for this test because pytest reads a list of numpy arrays as a list. Understand how pytest could iterate over each array and change code accordingly.

Reply:
The following seems to work for me:
``` python
@pytest.mark.parametrize("state",
    [np.array([[1,4,3]]),
        np.array([[1]]),
        np.array([[1, 2], [3, 4],[5,6]])
    ])
def test_normalize_matrix_non_square_matrix(state):
    """Test error is raised when row number and column number of Matrix
    are mismatched."""
    with pytest.raises(ValueError, match="Input must be a square matrix to be a valid gate."):
        normalize_matrix(state)
```

Notice that the test in the current version is actually not working in an ideal way. It will succeed if *any* of the call raises a `ValueError`, not *all* of them. And the array `np.array([[1]])` is actually not raising an error, which fails the test in the above test.

Reply:
Thanks ! I was actually creating a list of arrays via `()` instead of an array of arrays (like what you did). Most of the examples in [documentation](https://docs.pytest.org/en/6.2.x/parametrize.html) are using `[]` instead of `()`. I guess I just got lucky with my other examples. 😅